### ISSUE TITLE
NixOS in Windows Store for Windows Subsystem for Linux (WSL)

### ISSUE THREAD

**Issue by ThomasMader** [NixOS/nixpkgs#30391]
Nix already runs in Ubuntu via WSL afaik.
I think it would make sense to add an official NixOS image to the Windows Store.
Arch Linux is trying already. (https://github.com/Microsoft/BashOnWindows/issues/2253)

Are there any showstoppers?

**Comment by bencoman** [comment]
This would be super cool.  I'd like to move to using NixOS, but I can't give up my Windows machine, and I don't want to run it in a virtual machine.  What I'd *really* like to try doing is a continuous integration system using NixOS based build management to kick off compilation on the Windows side.

#28365 is an impediment, but drilling through the linked issue I saw an update yesterday from a MS staffer that they have been working on it.

Also I'm not sure if NixOS/nix#1203 is the same or different root issue.

**Comment by PierreR** [comment]
I have a permssion denied just after the creation of `/nix` when I try to install `nix` on WSL
![nix](https://user-images.githubusercontent.com/93115/32025618-4069ee6a-b9e1-11e7-88ae-5c7d18c411c5.PNG)

Any idea ?

**Comment by jtojnar** [comment]
@PierreR I also had this problem. I solved it by downloading the archive and copying the files to the directory manually.

**Comment by PierreR** [comment]
@jtojnar Thanks ! Indeed if I run each problematic `mv` myself, it is going to work.

Any idea where the issue is coming from. I would like to streamline that process but it is going to be a bit cumbersome if such issue lingers on.

**Comment by PierreR** [comment]
Just for the record I have hit https://github.com/Microsoft/WSL/issues/1671. Not only WSL seems to be unbearably slow, it would also fail strangely with this error:
```
error: setting synchronous mode: disk I/O error
```
My feeling is that WSL is too beta and adding an official NixOS image to the Windows Store might be premature.

**Comment by jtojnar** [comment]
@PierreR disk I/O error is https://github.com/NixOS/nix/issues/1203

**Comment by PierreR** [comment]
@jtojnar yes it is. Thanks ! Everything works fine for me except for the sluggishness.

**Comment by CMCDragonkai** [comment]
Does NixOS the distro itself require anything special beyond what Ubuntu as a distro requires? If not, then it should be possible to get NixOS (and not just nix) onto WSL!

**Comment by benley** [comment]
NixOS the distro relies pretty heavily on systemd to run services.  It would still be great to use `nixos-rebuild` for building the user environment and making software available, but without systemd it would be difficult to use any nixos modules that expect to run daemons or cronjobs.

**Comment by mdekstrand** [comment]
> NixOS the distro relies pretty heavily on systemd to run services. It would still be great to use nixos-rebuild for building the user environment and making software available, but without systemd it would be difficult to use any nixos modules that expect to run daemons or cronjobs.

Other distros will be in the same boat for any required services, but I can see this being a particular problem for Nix with the Nix Daemon.

Since WSL is intended to be a single-user environment, a NixOS that runs like single-user Nix could be what is needed to make it work.

**Comment by ThomasMader** [comment]
Microsoft open sourced a WSL Sample for Linux Distribution maintainers. Might be very helpful.
https://blogs.msdn.microsoft.com/commandline/2018/03/26/wsl-distro-launcher/

**Comment by DarthSpock** [comment]
Any news in this front? I'd manually compile myself to test it out but won't have time for it for awhile. If anybody on the NIXOS team is running build 17134 SCU (pre-release still), if they could give it a go on this build, it would be appreciated.

**Comment by ghuntley** [comment]
https://github.com/Microsoft/WSL-DistroLauncher/blob/master/README.md

https://docs.microsoft.com/en-us/windows/wsl/build-custom-distro

**Comment by ghuntley** [comment]
I'm taking a stab at this. Initial success so far @ https://twitter.com/GeoffreyHuntley/status/995977211176927232

**Comment by ghuntley** [comment]
Blocked by https://github.com/Microsoft/WSL/issues/3207

**Comment by PerArneng** [comment]
Thanks for the work on this. Testing out linux distros on Windows is really easy now days and you can run several side by side. Getting NixOS on to the app store could be a great gateway drug to NixOS and the excellent Nix package manager.

**Comment by ghuntley** [comment]
In private discussions with Microsoft about getting NixOS into the Microsoft App Store. Have been a little overwhelmed recently but it's high on my todo list. Microsoft are waiting on me.

**Comment by bencoman** [comment]
Thankyou Geoffrey. That would be great.  Good luck.
cheers -ben

**Comment by ghuntley** [comment]
Considering nixcon is on; it only makes sense to pick this back up again. I had to parked it because was overcommitted. Let's start this back up again.

<img width="1326" alt="screen shot 2018-10-25 at 9 18 01 pm" src="https://user-images.githubusercontent.com/127353/47493960-4b064500-d89c-11e8-96c0-b40a2cfb4cdd.png">

Next steps are as follows:
- [ ] NixOS Foundation to sign up for a  "Company" Windows Developer Account
- [ ] NixOS Foundation to review contracts and execute them
- [ ] Obtain testing plan from Microsoft
- [ ] Obtain publishing and release plan from Microsoft
- [ ] Productionise the NixOS build of `install.tar.gz` and embed into an appx.
- [ ] Distribute the appx to microsoft
- [ ] Party

**Comment by matthewbauer** [comment]
What's the structure of install.tar.gz? I couldn't find that documented.

**Comment by zimbatm** [comment]
@ghuntley do you have a URL for the NixOS Foundation to review the documents?

**Comment by hypersw** [comment]
Thanks for pushing NixOS towards running over WinNT! That would be handy.

The roadmap above is much about putting NixOS into the appstore. Which is quite separate a task from making NixOS runnable on WSL. I'd pretty much do with a ZIP which I'd download from nix site and install for me, not sure I even like the idea of mouseclicking it out of the fancy store instead. Especially if we have to wait for the legal stuff just for the sake of some GUI install scenario.

What's the technical status of running NixOS on WSL? Is it still blocked by the question marks and pipes in the paths? If not, are there any distributions we could try and run right away?

**Comment by matthewbauer** [comment]
> What's the technical status of running NixOS on WSL? Is it still blocked by the question marks and pipes in the paths? If not, are there any distributions we could try and run right away?

I don't have WSL, but in theory, it should be possible now using "containerTarball":

https://github.com/NixOS/nixpkgs/blob/master/nixos/release.nix#L217-L221

You can manipulate this into an install.tar.gz file and get it to run (and make a PR to get it in future release). There will most likely be bugs encountered doing this though. wslu looks to be a good resource to base the install.tar.gz on.

Another question that remains to be solved: how can we get Nix to detect that it can run Windows binaries? Nix supports "extra-platforms" so we can reuse that if there were an easy way to run a builder in "windows" mode. Can't find any sort of syscall/API to do this though.

**Comment by DarthSpock** [comment]
The proprietary init does the windows interop calls to the apis. Far as daemons go, as I stated before, you dont need systemd to run services. In Debian-based distros, you just do service blah start. For example ssh. Want it to start automatically? Not supported out of box yet but you can add to your bashrc, zshrc, etc. as workaround

**Comment by wmertens** [comment]
Now granted, WSL should support all POSIX filenames, and https://github.com/Microsoft/WSL/issues/3207 is definitely a bug on their side. However, would it not be relatively simple to replace all NTFS-invalid characters in Nix store path names with `_`?

Names clashing seems very unlikely since the input hash is prepended. There may be some derivations that have illegal characters in the package content filenames, those can be handled on a case-by-case basis.

Alternative is to wait for https://github.com/NixOS/rfcs/pull/17 ;-)

**Comment by ghuntley** [comment]
Current status:
- [still waiting on nixos foundation to sign legal paperwork to allow distribution into the app store]
- systemd issues resolved by today's announcement at build//
- following up with WSL product managers about getting access to WSL2 bits (under NDA if need be)

https://devblogs.microsoft.com/commandline/announcing-wsl-2/

**Comment by ghuntley** [comment]
Developer agreement is at https://docs.microsoft.com/en-us/legal/windows/agreements/app-developer-agreement

```
Hi Geoffrey and Graham,


Thanks for reaching out. It’s nice to virtually meet you. Here’s an overview of the process:

As outlined below, you’ve completed the first step by reaching out. I’ve pasted the other steps while adding additional details.

*  Agree with the WSL team on a testing and publishing plan 
*  We have some basic manual tests that we conduct such as testing networking. Are there any specific compatibility concerns you have for NixOS running on WSL?
*   The other part of publishing is that we ask you to send us your appx file. This will be put on a special Store CDN that will allow customers to install your app offline. You can read more on what that would look like 
*   Complete any required paperwork
*   When you sign up as a company for a Windows Developer Account you will need to sign an App Developer Agreement. Please take a look at that agreement and let us know if it meets your legal requirements. There are a couple other documents in the aforementioned link, so please take a look at those as well.
*  Sign up for a "Company" Windows Developer Account https://developer.microsoft.com/en-us/store/register. note: this can take a week or more since you'll be required to confirm your organization's identity with an independent verification service via email and/or telephone.

Once the above is complete, you can submit your app and stage it for publication. There are a couple things that happen in the background with the Store that can take a little while depending on the volume of submissions they have at the time.
```

**Comment by ghuntley** [comment]
Next steps in theory are:
- obtain WSL2 bits
- cook a nothing out of the ordinary virtual machine via CI
- wrap virtual machine with WSl2 bits and package as appx
- nixos foundation sign legal agreements and register for corporate developer account

**Comment by puffnfresh** [comment]
What does "ordinary" look like? Does that include systemd?

**Comment by ghuntley** [comment]
That's the missing piece of technical detail. Discussing now with Microsoft product managers. 

Not sure if we need to cook an ova or they take a tarball of a freshly installed system (WSL1) and then inject it into a VHD.

Not sure if they invoke grub when booting this VHD.

Not sure if initd/systemd is invoked or how to get WSL2 to invoke the equivalent.

**Comment by WSLUser** [comment]
There's no plan to implement systemd at this time. That's what was said at their MS Build booth when I asked. But init rc scripts work just fine. You can do that now without WSL2.

**Comment by matthewbauer** [comment]
Yeah, NixOS has an option to disable stage 1 with ```boot.isContainer```. We probably need to rework stage-2 a bit to avoid running systemd directly, but also keeping as much of the activation scripts as possible. The script is here: https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/system/boot/stage-2-init.sh. @cleverca22 has worked on some custom init scripts through not-os, which may be useful.

At the same time, providing an installer with just multi-user Nix installed would be super useful in its own right! We don't necessarily need to go full NixOS WSL for this to be useful. It sounds like we'll need to figure out how to automatically start the daemon correctly. Single user will probably work without any modifications!

**Comment by ghuntley** [comment]
Some more technical information can be found at https://devblogs.microsoft.com/commandline/shipping-a-linux-kernel-with-windows/ that's the best info that exists at this stage.

**Comment by ghuntley** [comment]
> It sounds like we'll need to figure out how to automatically start the daemon correctly.

I agree. 

> Single user will probably work without any modifications!

It does and it's what we should aim for.

https://twitter.com/GeoffreyHuntley/status/1007607578933714945

**Comment by tycho01** [comment]
They're releasing a [2.0](https://devblogs.microsoft.com/commandline/announcing-wsl-2/) but I presume that shouldn't affect making NixOS available there.

**Comment by edolstra** [comment]
@ghuntley To clarify, why does the foundation need to register as an App Developer? I assume you already have the ability to upload apps to the store?

**Comment by vatsalyagoel** [comment]
@edolstra It isn't recommended to ship an official distribution of an application under an individual's account. It brings other legal issues. Ownership of accounts and just general branding.

If he were to publish NixOS under his own account he would either have to give up ownership of his developer account to NixOS foundation or the ownership of NixOS distro on windows is handed over to Geoff

**Comment by edolstra** [comment]
Okay, I've registered an account for the NixOS Foundation. What's the next step?

**Comment by ghuntley** [comment]
WSL2 was released today and is now available for folks who are on the Windows 10 fast ring.

https://devblogs.microsoft.com/commandline/wsl-2-is-now-available-in-windows-insiders/

> Okay, I've registered an account for the NixOS Foundation (currently pending verification). I've also reserved "NixOS" as an app name.

Thank-you for registering the account and reserving the application name @edolstra.

> What's the next steps?

I'd love to team up with a couple folks who are more experienced with nixos internals and ship https://github.com/NixOS/nixpkgs/pull/63049 which was raised moments ago. Stop on by the pull-request, say howdy and I'll grant you push access to the pull-request branch.

**Comment by AndrewMorsillo** [comment]
It looks like there is a way to run systemd on WSL2 https://github.com/arkane-systems/genie although I'm not sure if that is useful

**Comment by ghuntley** [comment]
@AndrewMorsillo useful. Thanks for sharing, was unaware!

**Comment by razvan-panda** [comment]
@ghuntley any progress on this? Would you like to receive any help with it? Is your progress stored somewhere in a public repository?

**Comment by dustinlacewell** [comment]
I just came across this issue and am super excited to be able to run NixOS on WSL2!

**Comment by GuillaumeDesforges** [comment]
Same, really enthousiastic about NixOS on WSL2 !! Would be the best 😄

**Comment by thedavidmeister** [comment]
nix shell works OK on WSL 2 ubuntu FYI

**Comment by razvan-flavius-panda** [comment]
@thedavidmeister The problem is not about running Nix under WSL. That is obviously possible. You can also set up X11 to work with WSL and then you have something akin to a lightweight VM on Windows using Ubuntu.

This ticket is about having NixOS as an option for WSL. Basically NixOS should be the operating system being used by WSL and not Ubuntu or whatever else.

**Comment by thedavidmeister** [comment]
@razvan-flavius-panda i know what this issue is about but my experience with onboarding windows people to a nixos based project is that none of this is obvious so i'm trying to make it clearer what options are for people waiting on this to land :)

**Comment by razvan-flavius-panda** [comment]
Fair enough. Fair enough.

**Comment by zimbatm** [comment]
Just as a recap:

* WSL2 works more like a container than a real VM. You don't get to choose the linux kernel and there is no init system.
* NixOS is designed to run like a real linux distro, with a bootloader, selection of kernel and systemd as it's init system. Most NixOS modules depend on systemd to work properly.

I just wanted to communicate clearly that it's not just a small patch away.

From there, there are two routes possible:

1. The short and hacky route. It's possible to override the nixos modules to make it mostly work. This is what [arion](https://github.com/hercules-ci/arion) does for example. The main downside is that things will never be 100% compatible and prone to break during channel updates.
2. The proper route. This would require to go an split NixOS up properly. Define a common interface for cross-init systems. Mark modules that only works on "NixOS Linux" (vs "NixOS WSL2").

IMO option (2) is much better to avoid giving a false expectations. It's definitely a project but it's not completely insurmountable to do.

**Comment by WSLUser** [comment]
> You don't get to choose the linux kernel and there is no init system

You actually can import your own kernel for the past 2 or 3 Insider builds. There's a few other things as well. See the blogs. And there is an init system. There's just not systemd without genie. Create init.rc scripts and you'll have the same as having systemd. This would be a bit more challenging for several obvious reasons. But it's doable.

**Comment by WSLUser** [comment]
Also, NX-OS would still have to work on WSL1 as that is the default even after enabling WSL2.  But it's still the same as my last comment.

**Comment by jacereda** [comment]
> There's just not systemd without genie. Create init.rc scripts and you'll have the same as having systemd. This would be a bit more challenging for several obvious reasons. But it's doable.

This would be the way to go IMHO, this isn't specific to WSL:

https://github.com/NixOS/nixpkgs/issues/26067
https://github.com/NixOS/nixpkgs/issues/26850

**Comment by AndrewMorsillo** [comment]
I don't pretend to understand enough about linux and systemd to know if this is useful, but there's a long thread about systemd on WSL2 here that might have some useful information

https://github.com/microsoft/WSL/issues/994

**Comment by WSLUser** [comment]
I will note that I've seen plenty of init.rc scripts that including calling systemctl and other systemd functions (we have some for a few Fedora VMs/containers due to things used for managing the hosts.) So the functionality can remain exactly the same except you can add to the scripts other non-systemd stuff so if you don't have systemd, you don't run any of the systemd specific stuff. For redundancies found within a script, those should be abstracted away as much as possible to keep the file size down as well as possibly improve perf from not having to read so many lines. We do want to keep these scripts as small as possible following best coding practices.

**Comment by puffnfresh** [comment]
I looked into Genie and figured out how it's running systemd. I then wrote a NixOS module to wrap systemd and systemctl instead:

https://github.com/puffnfresh/nixpkgs/commit/08d8c23eebb0a5ab90db94a05e297630227cb232

It's somewhat working. Sometimes systemd starts up well and you can run systemctl to check on services, etc. It seems like `nixos-rebuild switch` mostly doesn't work, sadly. Still quite a bit of debugging to do.

**Comment by DanielFabian** [comment]
Whilst the `nixos-rebuild switch` is a bit of a shame, I think it's less important on WSL2 as the reboot time is only 1-2 seconds anyway.

**Comment by AndrewMorsillo** [comment]
Looks like somebody has it working 100% here, but I haven't tried it out yet https://github.com/Trundle/NixOS-WSL

**Comment by wmertens** [comment]
Systemd now has the --system flag which lets it run even if it doesn't have PID 1, so maybe all we need to do is set that flag?

**Comment by lulcat** [comment]
lol? I run a non systemd distro, and I have no issues with running systemd inside containers (non pid 1) fully non degraded, so didn't even systemd needed such flag? Out of curiousity, why on EARTH would anyone want / need to run linux on windows and not vice versa? Is it basically an update to cygwin or so? 2020? ye, 20 year late MS.. 90s os :) what is WSL?

**Comment by lulcat** [comment]
OH.. NixOS as I understood it was meant to be so 'smart about running several versions' but it can't even handle different inits? I understand one can blame systemd for that, but it's usually easier to run all good citizens together, as options, and leave the one bad one out. Still, I guess NixOS , despite having grown, isn't really all that then.

**Comment by wmertens** [comment]
@lulcat The init system used to be upstart iirc, feel free to find those commits and propose a maintainable combined system. There's also a lot to read in previous issues on this topic.

The gist of it is that it's hard to create a wrapper around init systems with such diverse capabilities without being very limiting or re-writing systemd capabilities; as such, NixOS modules use only systemd.

**Comment by GuillaumeDesforges** [comment]
Seems like it works now.
Any news on a possible release on the Microsoft Store ?

**Comment by WSLUser** [comment]
At the very least, an offline msix could be provided (that's still signed) to make usable outside of Store. The wsl --import option is great but an offline msix would be eve greater.

**Comment by WSLUser** [comment]
Not Stale - still not in the Store

**Comment by hypersw** [comment]
Are there any *specific* instructions for planting it into the WSL? Preferably, ones which do not involve writing an app in C++ and compiling it with Visual Studio. Everything there sounds not too hard — if you've done it before, and got all the tooling. But for the first time…

If not in the Store, a ready set of the DistroLauncher and `install.tar.gz` would be nice to have.

Rumors are, you can install an `install.tar.gz` by placing it alongside an `Ubuntu.exe` and running, but the part which says "It is expected that /nix/var/nix/profiles/system/activate is run during installation, for example as part of the DistributionInfo::CreateUser function" sounds as if it could be a problem.

Another issue is that you need NixOS to make a tar.gz you need to run NixOS, which in context of our task (running NixOS on Windows NT) sounds close to "unzip.zip".

**Comment by scalavision** [comment]
@hypersw , is this helpful maybe? If I have time, I will try it out during the weekend:

https://github.com/sonowz/vscode-remote-wsl-nixos

@GuillaumeDesforges or @WSLUser , do you know if anyone actually have tried to get this into Windows Store? Is there a request I could give thumbs up to somewhere? (don't know anything about how windows store works) ..

**Comment by GuillaumeDesforges** [comment]
> @GuillaumeDesforges or @WSLUser , do you know if anyone actually have tried to get this into Windows Store? Is there a request I could give thumbs up to somewhere? (don't know anything about how windows store works) ..

From @ghuntley (https://github.com/NixOS/nixpkgs/issues/30391#issuecomment-411413299)

> In private discussions with Microsoft about getting NixOS into the Microsoft App Store. Have been a little overwhelmed recently but it's high on my todo list. Microsoft are waiting on me.

Other than that idk.

It would help to have someone who knows the procedure to be accountable (maybe a strong word but I don't know another) on this (both follow technical issues and WS request)

**Comment by pecigonzalo** [comment]
> Seems like it works now.
> Any news on a possible release on the Microsoft Store ?

It would be great to have some manual steps to get it running as well. Is https://github.com/Trundle/NixOS-WSL the recommended way for example?

**Comment by allumik** [comment]
Is following the [Trundle's repo](https://github.com/Trundle/NixOS-WSL) still the best way to generate latest NixOS for WSL? Or are some forks or other repos that have 20.09+ images available? Possibly with WSLg, wslu support etc.

**Comment by alexvorobiev** [comment]
I [commented](https://github.com/Trundle/NixOS-WSL/issues/49#issue-1060671729) in that repo that [distrod](https://github.com/nullpo-head/wsl-distrod) claims to help with the systemd issue. Distrod relies on https://linuxcontainers.org which doesn't have anything for NixOS. Would it be difficult to create NixOS images for LXC/LXD and linuxcontainers.org so that distrod could be leveraged?

**Comment by asbachb** [comment]
I'm running NixOS on lxd, but tbh it's not straight forward and you need to configure quite a lot in os level. There's also https://github.com/nix-community/nixos-generators which is able to generate LCD images.

19 Jan 2022 19:29:32 Alex Vorobiev ***@***.***>:

**Comment by sielicki** [comment]
https://github.com/nix-community/NixOS-WSL/issues/63

Please close this and redirect discussion here.

**Comment by omasanori** [comment]
Closing as requested. Feel free to reopen if it is not appropriate.
